oximeter/db/src/native/io/packet/server.rs

```diff
@@ -0,0 +1,406 @@
+                        // next time.
+                        return Ok(None);
+                    }
+                    Err(e) => {
+                        // We failed to correctly decode the packet.
+                        //
+                        // We may need to do something else in the future, but
+                        // for now, just throw away the buffer. We have no idea
+                        // how much data we need to lop off. We should
+                        // realistically RST the whole TCP connection, most
+                        // likely.
+                        probes::invalid__packet!(|| ("Hello", src.len()));
+                        src.clear();
+                        return Err(e);
+                    }
+                }
+            }
+            Packet::DATA => match io::block::decode(&mut buf) {
+                Ok(Some(block)) => Packet::Data(block),
+                Ok(None) => return Ok(None),
+                Err(e) => {
+                    probes::invalid__packet!(|| ("Data", src.len()));
+                    src.clear();
+                    return Err(e);
+                }
+            },
+            Packet::EXCEPTION => match io::exception::decode(&mut buf) {
+                Ok(Some(exceptions)) => Packet::Exception(exceptions),
+                Ok(None) => return Ok(None),
+                Err(e) => {
+                    probes::invalid__packet!(|| ("Exception", src.len()));
+                    src.clear();
+                    return Err(e);
+                }
+            },
+            Packet::PROGRESS => {
+                let Some(progress) = io::progress::decode(&mut buf) else {
+                    return Ok(None);
+                };
+                Packet::Progress(progress)
+            }
+            Packet::PONG => Packet::Pong,
+            Packet::END_OF_STREAM => Packet::EndOfStream,
+            Packet::PROFILE_INFO => {
+                let Some(info) = io::profile_info::decode(&mut buf) else {
+                    return Ok(None);
+                };
+                Packet::ProfileInfo(info)
+            }
+            Packet::PROFILE_EVENTS => match io::block::decode(&mut buf) {
+                // Profile events are encoded as a data block.
+                Ok(Some(block)) => Packet::ProfileEvents(block),
+                Ok(None) => return Ok(None),
+                Err(e) => {
+                    probes::invalid__packet!(|| ("ProfileEvents", src.len()));
+                    src.clear();
+                    return Err(e);
+                }
+            },
+            _ => {
+                // We don't know anything about the packet, drop it.
+                //
+                // Again, we probably need to handle this more gracefully, but
+                // it's not clear how in the absence of a header with the packet
+                // size.
+                probes::unrecognized__server__packet!(|| (u64::from(kind), src.len()));
```

Comment:
I believe rustfmt would like this wrapped: https://github.com/oxidecomputer/omicron/actions/runs/10892927436/job/30226846221?pr=6584#step:5:9

Reply:
Yeah, I have that fixed. I'm still testing this a bit and trying to figure how to run `criterion` for some benchmarks...

Thanks for the early input!